### PR TITLE
Implement SIFT feature detector and descriptor

### LINKED ISSUE #4499: SIFT patent expired
I see on [HN](https://news.ycombinator.com/item?id=22519866) that SIFT patent is expiring. Any plan to implement it in skimage? :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by faberno**
## Description

Fixes #4499 
This PR implements the SIFT feature detection and its descriptor. It was originally developed by David Lowe (https://doi.org/10.1109/ICCV.1999.790410), but this implementation is is mostly based on the paper "Anatomy of the SIFT Method" and its demo by Rey Otero (https://doi.org/10.5201/ipol.2014.82).
I tried to mostly copy the style of ORB with the three Methods: detect, extract, detect_and_extract.
Basic tests and an example in ./doc/examples are provided.

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/feature/sift.py
Comment (mkcor) [inline skimage/feature/sift.py:89]:
```suggestion
              :DOI:`10.1109/ICCV.1999.790410`
```

File: skimage/feature/sift.py
Comment (mkcor) [inline skimage/feature/sift.py:94]:
```suggestion
              :DOI:`10.5201/ipol.2014.82`
```

File: skimage/feature/tests/test_sift.py
Comment (mkcor) [inline skimage/feature/tests/test_sift.py:15]: @grlee77 do we have a preference for writing `np.float32` over `'float32'`?

Comment (grlee77) [reply skimage/feature/tests/test_sift.py:15]: I don't have a strong preference either way. I think using strings may have made the parameterized test case names clearer with older pytest versions, but I'm pretty sure for > 1.5 years now it checks the `__name__` attribute of objects and uses that.

**Comment by grlee77** [comment]
Thanks @faberno. We are sprinting for scikit-image this weekend as part of SciPy 2021. Hopefully I will be able to review this sometime during that time. If you want to join in, say hello at https://skimage.zulipchat.com/#narrow/stream/176555-sprint/topic/scipy2021 or via the discord linked in the comments there.

**Comment by mkcor** [comment]
Thank you for your contribution, @faberno!

I'm reading your demo https://2338-2014929-gh.circle-artifacts.com/0/doc/build/html/auto_examples/features_detection/plot_sift.html#sphx-glr-auto-examples-features-detection-plot-sift-py and I can't really tell what is going on. It's definitely okay to be concise and link to external references but, as a reader, I don't understand what it is that we're trying to achieve here. When I see
original_image
I'm not sure what's going on, I'm not sure even what the 'original image' is (is supposed to look like). I understand that you have drawn inspiration from the ORB gallery example https://scikit-image.org/docs/stable/auto_examples/features_detection/plot_orb.html#sphx-glr-auto-examples-features-detection-plot-orb-py but, in this case, the matching lines are so many that it's technically impossible to view the underlying image, let alone its features.

**Comment by faberno** [comment]
@mkcor True, I didn't think of that. For the moment I changed it by only taking the first 100 keypoints and matching these. Would you prefer a n_keypoints parameter? Here we need to decide on which properties we select the keypoints:
- contrast
- edge response
- or combination of both?

A keypoint can also have multiple orientations and a copy is created for each one of these.
Should we rather take
- lots of different keypoints with only one orientation each
- less different keypoints with all its orientations

File: skimage/feature/tests/test_sift.py
Comment (grlee77) [inline skimage/feature/tests/test_sift.py:127]:
```suggestion
def test_no_descriptors_extracted_sift():
```

File: skimage/feature/tests/test_sift.py
Comment (grlee77) [inline skimage/feature/tests/test_sift.py:25]: What is the source of the expected values in the new tests? Are these obtained from a third-party reference implementation or were these obtained from the implementation here and intended just to protect against unexpected future changes in behavior?

File: doc/examples/features_detection/plot_sift.py
Comment (grlee77) [inline doc/examples/features_detection/plot_sift.py:26]: I think I would define a variable, something like `num_keep = 100` and add a comment above this noting that we have restricted the number of keypoints and matches for clarity in the plots

My only concern is that this could be misleading in terms of how many features SIFT typically finds. Maybe it is worth keeping the plot with all of the features first and then creating a second figure where the features are limited to a subset of 100.

File: skimage/feature/sift.py
Comment (grlee77) [inline skimage/feature/sift.py:173]: You don't have to address this comment immediately, but longer term we should think about how to refactor the scale space code in `skimage/feature/blob.py` so it could be reused here as well. I think we will need first the following fix there
https://github.com/scikit-image/scikit-image/pull/4482
and then I think there was another PR that had started that refactoring. I will link it here when I find it.

Comment (grlee77) [reply skimage/feature/sift.py:173]: It was this issue and corresponding PR. It is pretty old, though, so has become a bit out of sync with recent changes:
https://github.com/scikit-image/scikit-image/issues/2315
https://github.com/scikit-image/scikit-image/pull/2386

File: doc/examples/features_detection/plot_sift.py
Comment (faberno) [reply doc/examples/features_detection/plot_sift.py:26]: myplot
@grlee77 Would you prefer it like this?

File: skimage/feature/tests/test_sift.py
Comment (faberno) [reply skimage/feature/tests/test_sift.py:25]: The second. The values are obtained from the implementation.

**Comment by grlee77** [comment]
@faberno: 
While reviewing this I came up with many stylistic and performance-based modifications. Rather than try to add a ton of comments here, I have opened a PR for your review at: faberno/scikit-image#1

**Comment by grlee77** [comment]
Another publication by the same author as the IPOL article followed here did a further study of what parameters affect the stability of the keypoints:
https://arxiv.org/abs/1511.08478

Based on their conclusions, we could reduce the position refinement iterations from 5 to 2 without reducing stability. Other conclusions where that the `c_dog` threshold does not help much with filtering out unstable keypoints and that using a larger number of `n_scales` can improve stability. The estimate of the blur of the input image is also quite important to stability (in the publication above they refer to the method in this IPOL article: http://www.ipol.im/pub/art/2012/admm-nppsf/)

**Comment by faberno** [comment]
Besides the changes I mentioned[ at the end of your PR ](https://github.com/faberno/scikit-image/pull/1#issuecomment-902529970), I also changed:
- now self.n_octaves is the maximum number of octaves again, except if the image is too small. You removed this when you removed the warning in 7a296a94ecb. I'm not sure if it was intentional, but if it was I will remove this changes again
- the test and example outputs had to be changed again
- stylistic changes like formatting, removal of unused variables, etc

**Comment by faberno** [comment]
> Another publication by the same author as the IPOL article followed here did a further study of what parameters affect the stability of the keypoints:
> https://arxiv.org/abs/1511.08478
> 
> Based on their conclusions, we could reduce the position refinement iterations from 5 to 2 without reducing stability. Other conclusions where that the `c_dog` threshold does not help much with filtering out unstable keypoints and that using a larger number of `n_scales` can improve stability. The estimate of the blur of the input image is also quite important to stability (in the publication above they refer to the method in this IPOL article: http://www.ipol.im/pub/art/2012/admm-nppsf/)

That sounds very interesting! I will give it a look and do some tests.

File: skimage/feature/sift.py
Comment (grlee77) [inline skimage/feature/sift.py:7]:
```suggestion
from ..feature.util import DescriptorExtractor, FeatureDetector
```

File: skimage/feature/sift.py
Comment (grlee77) [inline skimage/feature/sift.py:11]:
```suggestion
from ._sift import (_local_max, _ori_distances, _oversample_bilin,
                    _update_histogram)
```

**Review by grlee77 — COMMENTED** [review]
The updates here look good to me! I made a few minor import style suggestions.

**Comment by grlee77** [comment]
Does the `_oversample_bilin` defined here not give the same result as using `rescale` with order=1?

File: skimage/feature/sift.py
Comment (grlee77) [inline skimage/feature/sift.py:270]:
```suggestion
        if self.upsampling > 1:
            image = np.ascontiguousarray(image)        
            image = _oversample_bilin(image, self.deltas[0])
```
Ensure the input to the Cython function is C-contiguous as expected.

**Review by grlee77 — CHANGES_REQUESTED** [review]

File: skimage/feature/sift.py
Comment (faberno) [reply skimage/feature/sift.py:270]: Sorry, forgot that.

**Comment by faberno** [comment]
> Does the `_oversample_bilin` defined here not give the same result as using `rescale` with order=1?

Thats how the difference between rescale and oversample_bilin looks like:
diff_rescale


But looks like with rescale more keypoints are found:

comparison

**Comment by grlee77** [comment]
The resampling coordinate conventions have been a common source of questions/confusion both here and in SciPy's ndimage. I made an illustration based on a 1D input signal with only 4 samples, that shows the difference between the upsampled coordinate locations used by `rescale` and `oversample_bilin`. It can be seen that there is an offset between the two. `scipy.ndimage.zoom` with grid_mode=False is different again from either of these two.

I would argue that the existing `resize` behavior is, in general desirable, for images. So perhaps, we should keep that part as it was rather than trying to exactly match the IPOL implementation on this aspect?

interp_coords

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:113]: Isn't this equivalent?
```suggestion
                w2 = w0 - w1
```

File: skimage/feature/sift.py
Comment (rfezzani) [inline skimage/feature/sift.py:635]: We can avoid a copy here :wink:
```suggestion
                histograms = histograms.reshape(-1)
```

File: skimage/feature/sift.py
Comment (rfezzani) [inline skimage/feature/sift.py:660]: These lines are repeated in `extract` and `detect_and_extract`. They can be refactored in a `_preprocess` function for example.

**Review by rfezzani — COMMENTED** [review]
Thank you @faberno, I just left some suggestions :wink:

**Comment by faberno** [comment]
> I would argue that the existing resize behavior is, in general desirable, for images. So perhaps, we should keep that part as it was rather than trying to exactly match the IPOL implementation on this aspect?

Thanks for the explanation! I now changed it back to rescale. For now I've not removed the function oversample_bilin in case somebody wants to recreate the results of the original demo

File: skimage/feature/_sift.pyx
Comment (faberno) [reply skimage/feature/_sift.pyx]: True, well spotted!

File: skimage/feature/sift.py
Comment (faberno) [reply skimage/feature/sift.py:241]: Good idea, but should we then make "dtype" a member of  the SIFT class, so we don't have to give it as a parameter to self.deltas?

Comment (rfezzani) [reply skimage/feature/sift.py:241]: >  should we then make "dtype" a member of the SIFT class

That's in fact a good option :wink:

Comment (faberno) [reply skimage/feature/sift.py]: Alright, I've added everything

File: skimage/feature/sift.py
Comment (faberno) [reply skimage/feature/sift.py]: Done!

**Review by grlee77 — APPROVED** [review]
I think this is in good shape.

We probably do want a commit to remove `_oversample_bilin`, but we can preserve it in the commit history. Lately, we typically squash all commits, but this seems like a case where it could be worth doing a manual rebase to squash a lot of the smaller ones, but preserve some important changes/revisions. I am happy to help with that rebase once we have another approval here.

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:101]:
```suggestion
    for p, (w, t_val, k_index) in enumerate(zip(magnitude, near_t_val, near_t)):
```

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:110]:
```suggestion
                      * w)
```

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:95]:
```suggestion
```

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:98]:
```suggestion
        np_floats t_val, val_norm1, w, w0, w1, w2, inv_spacing
```

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:111]:
```suggestion
```

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:114]:
```suggestion
```

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:118]:
```suggestion
                k_index2 = k_index + 1
                if k_index2 == n_ori:
                    k_index2 = 0
```

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:139]:
```suggestion
    for r in range(-1, 2):
        if r == 0:
            continue
        for c in range(-1, 2):
            if c == 0:
                continue
            for s in range(-1, 2):
                if s == 0:
                    continue
                neighbor_offsets[n] = (r * n_c + c) * n_s + s
                n += 1
```

**Review by rfezzani — COMMENTED** [review]
I added some suggestions, I will come with another round of review soon :wink:. Thank you again @faberno

File: skimage/feature/_sift.pyx
Comment (faberno) [reply skimage/feature/_sift.pyx:139]: I think that change wouldn't work out, because we would skip all 0's, but we actually just want to skip if r, c, and s all are 0.
But if we want to make the if a bit prettier, we could write
`if any([r, c, s]):`

Comment (rfezzani) [reply skimage/feature/_sift.pyx:139]: :confounded: I confused the `and` and `or` statement, you are right!

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:166]: I think that we can simplify this block:
```suggestion
                for offset in neighbor_offsets:
                    val = center_ptr[offset]
                    if val <= center_val:
                        for offset in neighbor_offsets:
                            val = center_ptr[offset]
                            if val >= center_val:
                                break
                        else:
                            # It is a local max
                            maxima_coords.append((r, c, s))
                    break
                else:
                     # It is a local min
                    maxima_coords.append((r, c, s))
```

File: skimage/feature/_sift.pyx
Comment (faberno) [reply skimage/feature/_sift.pyx:49]: I just tried that and it seems like the enumerate makes the function a lot slower

File: skimage/feature/_sift.pyx
Comment (rfezzani) [inline skimage/feature/_sift.pyx:36]: Shouldn't these `double` be `np_floats`?

File: skimage/feature/_sift.pyx
Comment (faberno) [reply skimage/feature/_sift.pyx:165]: Right now, this change doesn't work, at least the tests are failing

File: skimage/feature/_sift.pyx
Comment (faberno) [reply skimage/feature/_sift.pyx:36]: I agree, but it's probably more a question for @grlee77

File: skimage/feature/_sift.pyx
Comment (rfezzani) [reply skimage/feature/_sift.pyx:165]: Well, I am not sure to understand why :confused:. This refactoring should be equivalent
```suggestion
                is_local_min = True
                is_local_max = False
                for offset in neighbor_offsets:
                    val = center_ptr[offset]
                    if val <= center_val:
                        is_local_max = True
                        is_local_min = False
                        for offset in neighbor_offsets:
                            val = center_ptr[offset]
                            if val >= center_val:
                                is_local_max = False
                                break
                        break
                if is_local_min or is_local_max:
                    maxima_coords.append((r, c, s))
```

Comment (rfezzani) [reply skimage/feature/_sift.pyx:165]: @faberno

> Right now, this change doesn't work, at least the tests are failing

just to be sure, the `else` statements must be aligned with the `for` bloc, not the `if` bloc ^^

**Comment by rfezzani** [comment]
Fixes #4499

File: skimage/feature/sift.py
Comment (stefanv) [inline skimage/feature/sift.py:93]: Incorrect indentation

File: doc/examples/features_detection/plot_sift.py
Comment (stefanv) [inline doc/examples/features_detection/plot_sift.py:8]: It would be nice to have some context here about what SIFT is, how it is used, and what other feature detectors exist in skimage.

File: skimage/feature/sift.py
Comment (stefanv) [inline skimage/feature/sift.py:167]: Could you add a mention that it was patented, but that the patent expired?

https://patents.google.com/patent/US6711293B1/en

**Review by stefanv — COMMENTED** [review]
An impressive piece of work, @faberno, thank you!  It is exciting to see SIFT in skimage after all these years.

File: skimage/feature/_sift.pyx
Comment (faberno) [reply skimage/feature/_sift.pyx:165]: The second version would work. I'm a bit tired so correct me if I'm wrong: Aren't you checking the neighbors unnecessarily often by nesting the loops?
I guess we could put everything in one loop if we wanted to, but to me the initial solution looks quite good and intuitive :D

Comment (rfezzani) [reply skimage/feature/_sift.pyx:165]: > Aren't you checking the neighbors unnecessarily often by nesting the loops?

I don't think so, the `break` statements stops the search whenever it is done...

> I guess we could put everything in one loop if we wanted to, but to me the initial solution looks quite good and intuitive :D

By putting everything in one loop, I just want to avoid an `if ... else ...` bloc when looping over all the elements of the `octave` array...

Comment (faberno) [reply skimage/feature/_sift.pyx:165]: You're right, your version is a bit faster, I will add it :D

Comment (faberno) [reply skimage/feature/_sift.pyx:165]: Ahh, you're first version didn't work because the second break was wrongly indented. But I tested it and the second version with the booleans seems to be a tiny bit faster

Comment (rfezzani) [reply skimage/feature/_sift.pyx:165]: OK, thank you @faberno for your invetigations :wink:

File: skimage/feature/_sift.pyx
Comment (grlee77) [reply skimage/feature/_sift.pyx]: I fixed this, thanks. I think I initially started with just double and missed these when switching over to `np_floats`

File: skimage/feature/sift.py
Comment (rfezzani) [inline skimage/feature/sift.py:260]: Shouldn't `size_min` be a parameter here?

Comment (faberno) [reply skimage/feature/sift.py:260]: I thought about that too. In the original implementation this is a constant and can't be changed. My reason not to do it was that the user can still practically increase size_min by lowering n_octaves and I'm not sure if it makes sense to lower size_min even more.
But if you prefer it as a parameter I have no problem with that. We just need to find some minimal value, because size_min too low would probably start to break some things.

Comment (rfezzani) [reply skimage/feature/sift.py:260]: An other option would be to make it private attribute of the `SIFT` class to make it twickable for "expert" users...

Comment (faberno) [reply skimage/feature/sift.py:260]: That sounds good, I will add it.

Comment (grlee77) [reply skimage/feature/sift.py:260]: I plan to merge this PR soon for inclusion in the 0.19 release. If this does not get addressed before then, it can go in a follow-up PR as it would not be a public API change.

Comment (stefanv) [reply skimage/feature/sift.py:260]: Let's leave it as is for now.  We can always add this feature if it is requested, but if it is not a parameter in the original it seems unlikely.  Also, I am not enthusiastic about private attributes tweakable by expert users: who should either have an API for it, or it should not be adjustable.

**Comment by grlee77** [comment]
Okay, this is now in.  Thank you, @faberno for the large amount of work and your patience seeing this through!

**Comment by faberno** [comment]
Thanks for all of your support! It was a lot of fun and in the future I'll be happy to contribute again.

**Comment by stefanv** [comment]
It is fantastic to see SIFT finally implemented in scikit-image.  Thanks very much @faberno for this great contribution!

**Pull request merged**
